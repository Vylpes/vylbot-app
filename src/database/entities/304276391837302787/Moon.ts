import { Column, Entity, IsNull } from "typeorm";
import BaseEntity from "../../../contracts/BaseEntity";
import AppDataSource from "../../dataSources/appDataSource";

@Entity()
export default class Moon extends BaseEntity {
    constructor(moonNumber: number, description: string, userId: string) {
        super();

        this.MoonNumber = moonNumber;
        this.Description = description;
        this.UserId = userId;
    }

    @Column()
    MoonNumber: number;

    @Column()
    Description: string;

    @Column({ nullable: true })
    WhenArchived?: Date;

    @Column()
    UserId: string;

    public static async FetchMoonsByUserId(userId: string): Promise<Moon[] | null> {
        const repository = AppDataSource.getRepository(Moon);

        const all = await repository.find({ where: { UserId: userId } });

        return all;
    }

    public static async FetchPaginatedMoonsByUserId(userId: string, pageLength: number, page: number): Promise<[ Moon[], number ]> {
        const rangeStart = page * pageLength;

        const repository = AppDataSource.getRepository(Moon);

        const moons = await repository.findAndCount({
            where: { UserId: userId, WhenArchived: IsNull() },
            order: { MoonNumber: "ASC" },
            skip: rangeStart,
            take: pageLength,
        });

        return moons;
    }

    public static async FetchMoonCountByUserId(userId: string): Promise<number> {
        const repository = AppDataSource.getRepository(Moon);

        const count = await repository.count({ where: { UserId: userId } });

        return count;
    }
}
