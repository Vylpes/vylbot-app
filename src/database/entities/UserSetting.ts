import { Column, Entity} from "typeorm";
import AppDataSource from "../dataSources/appDataSource";
import BaseEntity from "../../contracts/BaseEntity";

@Entity()
export default class UserSetting extends BaseEntity {
    constructor(userId: string, key: string, value: string) {
        super();

        this.UserId = userId;
        this.Key = key;
        this.Value = value;
    }

    @Column()
    UserId: string;

    @Column()
    Key: string;

    @Column()
    Value: string;

    public UpdateValue(value: string) {
        this.Value = value;
    }

    public static async FetchOneByKey(userId: string, key: string, relations?: string[]): Promise<UserSetting | null> {
        const repository = AppDataSource.getRepository(UserSetting);

        const single = await repository.findOne({ where: { UserId: userId, Key: key }, relations: relations || {} });

        return single;
    }
}
