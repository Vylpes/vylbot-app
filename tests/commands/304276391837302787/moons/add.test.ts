import { CommandInteraction, EmbedBuilder } from "discord.js";
import AddMoon from "../../../../src/commands/304276391837302787/moons/add";
import Moon from "../../../../src/database/entities/304276391837302787/Moon";
import UserSetting from "../../../../src/database/entities/UserSetting";

describe("GIVEN happy flow", () => {
    let repliedWithEmbed: EmbedBuilder[] | undefined;
    let savedMoon: Moon | undefined;

    const interaction = {
        reply: jest.fn((options: any) => {
            repliedWithEmbed = options.embeds;
        }),
        options: {
            get: jest.fn()
                .mockReturnValueOnce({
                    value: "Test Description",
                }),
        },
        user: {
            id: "userId",
        },
    } as unknown as CommandInteraction;

    const userSetting = {
        Value: 1,
        UpdateValue: jest.fn(),
        Save: jest.fn(),
    };

    beforeAll(async () => {
        Moon.FetchMoonCountByUserId = jest.fn().mockResolvedValue(1);
        Moon.prototype.Save = jest.fn().mockImplementation((_, entity: Moon) => {
            savedMoon = entity;
        });

        UserSetting.FetchOneByKey = jest.fn().mockResolvedValue(userSetting);

        await AddMoon(interaction);
    });

    test("EXPECT description option to have been fetched", () => {
        expect(interaction.options.get).toHaveBeenCalledTimes(1);
        expect(interaction.options.get).toHaveBeenCalledWith("description", true);
    });

    test("EXPECT UserSetting to have been fetched", () => {
        expect(UserSetting.FetchOneByKey).toHaveBeenCalledTimes(1);
        expect(UserSetting.FetchOneByKey).toHaveBeenCalledWith("userId", "moons");
    });

    test("EXPECT moonCount to be updated +1", () => {
        expect(userSetting.UpdateValue).toHaveBeenCalledTimes(1);
        expect(userSetting.UpdateValue).toHaveBeenCalledWith("2");
    });

    test("EXPECT setting to be saved", () => {
        expect(userSetting.Save).toHaveBeenCalledTimes(1);
        expect(userSetting.Save).toHaveBeenCalledWith(UserSetting, userSetting);
    });

    test("EXPECT moon to be saved", () => {
        expect(Moon.prototype.Save).toHaveBeenCalledTimes(1);
        expect(Moon.prototype.Save).toHaveBeenCalledWith(Moon, expect.any(Moon));

        expect(savedMoon).toBeDefined();
        expect(savedMoon).toMatchSnapshot({
            Id: expect.any(String),
            WhenCreated: expect.any(Date),
            WhenUpdated: expect.any(Date),
        });
    });

    test("EXPECT embed to be replied", () => {
        expect(interaction.reply).toHaveBeenCalledTimes(1);

        expect(repliedWithEmbed).toBeDefined();
        expect(repliedWithEmbed).toMatchSnapshot();
    });
});

describe("GIVEN description is null", () => {
    test.todo("EXPECT error replied");
});

describe("GIVEN description is greater than 255 characters", () => {
    test.todo("EXPECT error replied");
});

describe("GIVEN moon count setting exists", () => {
    test.todo("EXPECT existing entity to be updated");
});

describe("GIVEN moon count setting does not exist", () => {
    test.todo("EXPECT new entity to be created");
});