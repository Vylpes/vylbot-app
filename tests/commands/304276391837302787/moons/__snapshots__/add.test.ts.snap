// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GIVEN happy flow EXPECT embed to be replied 1`] = `
[
  {
    "color": 5294200,
    "description": "**2 -** Test Description",
    "thumbnail": {
      "url": "https://cdn.discordapp.com/emojis/374131312182689793.webp?size=96&quality=lossless",
    },
    "title": "undefined Got A Moon!",
  },
]
`;

exports[`GIVEN happy flow EXPECT moon to be saved 1`] = `
{
  "Description": "Test Description",
  "Id": Any<String>,
  "MoonNumber": 2,
  "UserId": "userId",
  "WhenCreated": Any<Date>,
  "WhenUpdated": Any<Date>,
}
`;
